### ISSUE TITLE
UI: Should decimals be displayed as percents? 

### ISSUE THREAD

**Issue by MattHJensen** [OpenSourcePolicyCenter/webapp-public#269]
Currently Tax-Calculator and TaxBrain accept rates as fractions: If a tax rate should be 35%, the user enters 0.35. We recently had a discussion about how a user could enter "2" expecting 2% and instead getting 200%  https://github.com/open-source-economics/Tax-Calculator/pull/788#discussion-diff-68937111. 

To alleviate this potential for user error, should TaxBrain accept rates as percents? If you want 35%, you would enter "35%" in TaxBrain. This would be applied to haircuts, ceilings, floors, tax and phaseout rates, etc. 

One advantage is that it would be impossible to misinterpret what the user wants if we demand the percent sign. Another advantage that springs to mind is that TaxBrain would likely appear more friendly. I expect many users are more comfortable thinking about a 35% rate as 35% rather than .35. 

A disadvantage is that there would be a format mismatch between Tax-Calculator `current_law_policy.json` and TaxBrain. 

Looking for everyone's feedback here. CC @feenberg, @martinholmer, @zrisher because of their participation in the discussion cited above.

**Comment by feenberg** [comment]
On Thu, 30 Jun 2016, Matt Jensen wrote:

> 
> Currently Tax-Calculator and TaxBrain accept rates as fractions: If a tax rate should be 35%, the user enters
> 0.35. We recently had a discussion about how a user could enter "2" expecting 2% and instead getting 200%
> open-source-economics/Tax-Calculator#788 (diff).

Probably the website should pop up a warning for fractions greater than 1.

> 
> To alleviate this potential for user error, should TaxBrain accept rates as percents? If you want 35%, you
> would enter "35%" in TaxBrain. This would be applied to haircuts, ceilings, floors, tax and phaseout rates,
> etc.
> 
> One advantage is that it would be impossible to misinterpret what the user wants if we demand the percent sign.
> Another advantage that springs to mind is that TaxBrain would likely appear more friendly. I expect many users
> are more comfortable thinking about a 35% rate as 35% rather than .35.
> 
> A disadvantage is that there would be a format mismatch between Tax-Calculator current_law_policy.json and
> TaxBrain.
> 
> Looking for everyone's feedback here. CC @feenberg, @martinholmer, @zrisher because of their participation in
> the discussion cited above.

I think doing more than a warning is a diversion of resources from more 
vital tasks.

dan

**Comment by talumbau** [comment]
My main concern here would be to make sure that if someone enters `35%`, the code did not silently accept this as `35`. I don't think that is what happens right now. We can add code to make `35%` from the user into `0.35` pretty easily.

**Comment by zrisher** [comment]
My recommendation would be to just run some simple javascript on the field to display it as a percent but store it in the form field as a decimal. So all the server-side code would stay the same and deal with  it as a decimal, but to the user it would *appear* as a percent.

An example case, where the parameter has a default of zero:
* User loads the page, the field is displayed as `0%` and stored as `0`.
* User enters `5` in the field, % sign stays fixed on the right as they type, ends up displayed as `5%` and stored as `.05`.
* User enters `200`, displayed as `200 %` and stored as `2`.
* User enters `150%`, non-numeric characters are parsed out and it's displayed as `150%` stored as `1.5`
* User enters `abc`, non-numeric characters are parsed out and the consequently empty field is stored as the default value of `0`, displayed as `0%`

An additional advantage of restricting the input markup to web GUI is that it would keep the form field interface truer to taxcalc format for anyone that wants to make an API request as an HTTP post.

**Comment by zrisher** [comment]
> I think doing more than a warning is a diversion of resources from more vital tasks.

If this is the case, you could always do the warning version first and add the parsing JS down the line when prioritization permits. They both start with a JS function that targets that form field.

**Comment by feenberg** [comment]
On Thu, 30 Jun 2016, zrisher wrote:

> 
> My recommendation would be to just run some simple javascript on the field to display it as a percent but store
> it in the form field as a decimal. So all the server-side code would stay the same and deal with it as a
> decimal, but to the user it would appear as a percent.
> 
> An example case, where the parameter has a default of zero:
>
>  *  User loads the page, the field is displayed as 0% and stored as 0.
>  *  User enters 5 in the field, % sign stays fixed on the right as they type, ends up displayed as 5% and
>     stored as .05.
>  *  User enters 200, displayed as 200 % and stored as 2.
>  *  User enters 150%, non-numeric characters are parsed out and it's displayed as 150% stored as 1.5
>  *  User enters abc, non-numeric characters are parsed out and the consequently empty field is stored as the
>     default value of 0, displayed as 0%
> 
> An additional advantage of restricting the input markup to web GUI is that it would keep the form field
> interface truer to taxcalc format for anyone that wants to make an API request as an HTTP post.

But anyone used to the GUI will be surprised by the other input methods. 
I don't want to forsake fractions, only warn users of percents.

dan

**Comment by zrisher** [comment]
> But anyone used to the GUI will be surprised by the other input methods.
> I don't want to forsake fractions, only warn users of percents.

In my experience, web application users generally welcome and quickly adapt to interface changes that improve usability. But I don't have any empirical evidence for that which I can share. Happy to help with this in whatever format you guys expect is best for your particular set of users. @MattHJensen

**Comment by MattHJensen** [comment]
After giving this more thought over the last year, I am a proponent of formatting fractions as percents. In particular, I think it would make sense to follow @zrisher's suggestion from this thorough [comment](https://github.com/OpenSourcePolicyCenter/PolicyBrain/issues/269#issuecomment-229760165):

> My recommendation would be to just run some simple javascript on the field to display it as a percent but store it in the form field as a decimal. So all the server-side code would stay the same and deal with it as a decimal, but to the user it would appear as a percent.

@hdoupe @GoFroggyRun, what do you think about this?

 @GoFroggyRun, how do you perceive the difficulty?

**Comment by martinholmer** [comment]
@MattHJensen said recently in issue #269:
> After giving this more thought over the last year, I am a proponent of formatting fractions as percents. In particular, I think it would make sense to follow @zrisher's suggestion from this thorough comment:
>
>> My recommendation would be to just run some simple javascript on the field to display it as a percent but store it in the form field as a decimal. So all the server-side code would stay the same and deal with it as a decimal, but to the user it would appear as a percent.

But @MattHJensen didn't mention again his though on the same issue from over a year ago:
> A disadvantage is that there would be a format mismatch between Tax-Calculator `current_law_policy.json` [documentation, which appears automatically in TaxBrain] and [the way] TaxBrain [would show the value of the fraction].

This mismatch is a big issue.

But more significantly, how important is this seemingly time-consuming change in comparison to getting TaxBrain using CPS data and then getting Tax-Calculator and TaxBrain to use benefit data along with the CPS data?  I thought the latter objectives had a pretty high priority given looming grant deadlines.  I can't imaging the fraction-vs-percent issue can be more important than those given that nobody has discussed it in 16 months.

After PolicyBrain 1.1.0 is available this Friday, October 27, it would seem to me the top priority would be to get TaxBrain working with taxcalc 0.12.0, which means using basic CPS data as well as PUF data, plus a number of other basic changes (like making available on TaxBrain the recent user request on specifying new child tax credit reforms).

What about deferring consideration of issue #269 until TaxBrain is using CPS and CPS-benefit data and doing UBI analysis?

@MattHJensen @hdoupe @GoFroggyRun

**Comment by GoFroggyRun** [comment]
@MattHJensen said:

>After giving this more thought over the last year, I am a proponent of formatting fractions as percents. In particular, I think it would make sense to follow @zrisher's suggestion from this thorough comment:
>>My recommendation would be to just run some simple javascript on the field to display it as a percent but store it in the form field as a decimal. So all the server-side code would stay the same and deal with it as a decimal, but to the user it would appear as a percent.

I agree that percent display might be more intuitive to users comparing to fractions. The percentage representation, more importantly, is more user-friendly for that it requires less tedious input. Making those input displayed in a percent on front-end, while stored, still, as fractions on back-end, is not technically hard per se. However, I am hesitated that this change might confound the users. What we currently have in `current_law_policy.json`, as well as the format suggested by JSON file upload, are both in fractions. It would be more desirable to keep the formats of GUI input/display, JSON upload, and parameters in `current_law_policy.json` consistent. I am thus in line with @martinholmer that we might want to leave the current display format "as is" for the moment.

**Comment by martinholmer** [comment]
@GoFroggyRun said in part:
> What we currently have in `current_law_policy.json`, as well as the format suggested by JSON file upload, are both in fractions [rather than percents]. 

Excellent point!   I hadn't thought of that additional and **permanent** problem with file upload and TaxBrain GUI input being different.

**Comment by hdoupe** [comment]
I agree with @martinholmer that this adds an additional level of complexity that we don't have the band-with to handle right now.

@GoFroggyRun makes a good point about the mismatch of inputs across the access points for the models.  For users who use these models via direct Python calls and the webapp interface, this could be confusing.

**Comment by MattHJensen** [comment]
Thanks for the input everyone. No need to think about this more for now.

**Comment by hdoupe** [comment]
@MattHJensen Can we close #269?
